text/2021-10-08-monitoring-metrics-about-chaos-mesh.md

```diff
@@ -0,0 +1,241 @@
+# Monitoring Metrics about Chaos Mesh
+
+## Summary
+
+More metrics is needed for improving observability of `chaos-controller-manager`,
+`chaos-daemon` and `chaos-dashboard`. Now, we already have several metrics in
+`chaos-controller-manager` and `chaos-daemon`, but it's not enough.
+
+## Motivation
+
+At present, we only collected several metrics about webhooks like
+`chaos_mesh_injections_total`, chaos experiments information like
+`chaos-controller-manager`, and HTTP and gRPC in `chaos-daemon`.
+These metrics are not easy to reflect the overall appearance of chaos-mesh system,
+so we need more metrics to improve the observability.
+
+As for the expected outcome, below is several metrics about logic pattern and
+performance should be implemented and exposed on `/metrics` HTTP endpoint:
+
+chaos-controller-manager:
+
+- time histogram for each `Reconcile()` in `Reconciler`
+- count of Chaos Experiments, Schedule, and Workflow
+- count of the emitted event by chaos-controller-manager
+- common metrics of grpc-client
+- metrics for kubernetes webhook
+
+chaos-daemon:
+
+- common metrics of grpc-server (already provided)
+- count of processes controlled by `bpm` (background process manager)
+- count of iptables/ipset/tc rules
+
+chaos-dashboard:
+
+- time histogram for each HTTP query
+- count for archived object
+- time histogram for archive reconciler
+
+## Detailed design
+
+### Metrics Plan
+
+The metrics design is as follows. It should be noted that the implementations
+of `chaos_controller_manager_chaos_experiments_count` and
+`chaos_daemon_grpc_server_handling_seconds` have been provided, and they are
+called `chaos_mesh_experiments` and `grpc_server_handling_seconds`. It is just
+hoping to modify the names to standardize naming. More about naming, see this
+guideline: [Metric and label naming](https://prometheus.io/docs/practices/naming/).
+
+<!-- markdownlint-disable MD013 -->
+| Name                                                  | Description                                            | Type         | Label                       | Buckets                      |
+|-------------------------------------------------------|--------------------------------------------------------|--------------|-----------------------------|------------------------------|
+| chaos_controller_manager_reconcile_duration_seconds   | time histogram for each `Reconcile()` in `Reconciler`  | HistogramVec | type                        | DefBuckets                   |
```

Comment:
Just curious, are these `reconcilation` and `webhook` related metrics already exposed by `controller-runtime` library?
I am not 100% sure, just double-checking.

Reply:
Sorry, I didn't check it before. After running some chaos experiments, I saw that there are indeed `controller_runtime_webhook_latency_seconds{webhook}`, `controller_runtime_webhook_requests_total{code, webhook}` and `controller_runtime_webhook_requests_in_flight{webhook}` metrics exposed on the `/metrics` endpoint (source code: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/internal/metrics/metrics.go), but the metrics related to reconcile do not exist. 
Should we remain the new metrics `chaos_controller_manager_webhook_duration_seconds{type, allowed}` and `chaos_controller_manager_webhook_request_total{type, allowed}`? Only the metrics labels are different.

Reply:
I prefer to use metrics provided by controller-runtime. :)